### ISSUE TITLE
Apex class INT_InteractionProcessor error- issue with Mailing City Mapping

### ISSUE THREAD

**Issue by robertwynter** [SFDO-Community/Interactions-for-Student-Recruitment#21]
Using a Webform (FormAssembly) to create an Interaction & Getting an error. Using the same form multiple times. it sometimes errors at the first submit and sometimes after the third submission, I receive an error. I have a form that can be resubmitted multiple times if the subscriber wishes to update information being submitted. It works perfectly for the first two submissions from the same subscriber and by the third, it fails. Which doesn't make sense? I could add a new subscriber via the same form and it's the same thing, the new subscriber will error out on the third one.

I've reached out to FormAssembly Support and they said the error regards a trigger in Salesforce so it's out of their hands.

Error Message: 9. Create Interaction__c - INT_Interaction: execution of AfterInsert caused by: System.TypeException: Invalid conversion from runtime type Account to Contact Class.INT_InteractionProcessor.convertLeads: line 451, column 1 Class.INT_InteractionProcessor.pr

the error is occurring because of the trigger associated with Apex class "INT_InteractionProcessor" which is associated with HEDA

For the Create Interaction mapping, I removed everything except Name and email. Then it was a success. Started adding them back in one by one and when it came to mapping Mailing City. The error occurs.

This is telling that name and billingCIty is matching and standard contact matching rules are triggered to stop it..so it is not converting the lead but there is no need for lead ..it should not create a lead in this case. when I created manually with interactions it is working fine.

public class INT_InteractionProcessor { private List dupeInteractions = new List(); private List leadsToDelete = new List(); private List opportunitiesToUpsert = new List(); private Map interactionIdToLead = new Map(); private Map interactionMap = new Map(); private Map leadIdToInteractionMap = new Map(); private Set contactIds = new Set(); private Set existingLeadIds = new Set();

**Comment by robertwynter** [comment]
this for loop is what's causing the issue;
 for (Datacloud.MatchRecord match : matchResult.getMatchRecords()) {
                                // We want to use the oldest matched record, it will be the source Contact.
                                Contact compareContact = (Contact) match.getRecord();
                                if (matchedContact == null || (compareContact.CreatedDate < matchedContact.CreatedDate)) {
                                    matchedContact = new Contact();
                                    matchedContact = compareContact;
                                }
                            }

**Comment by robertwynter** [comment]
- Using V42 Spring 18

I checked the Matching configuration. The only difference was in the criteria filters on the number of Brackets. The documentation has the wrong number of brackets. We either had to add or remove a bracket for it to be correct.

I opened a support ticket and they came back with:

The error is appeared from the following code statement convertLeads(newLeads) from where the explicit dynamic typeCasting is done

for (Datacloud.MatchRecord match : matchResult.getMatchRecords()) {
// We want to use the oldest matched record, it will be the source Contact.
Contact compareContact = (Contact) match.getRecord()


Here match.getRecord is fetching the account and since the account and contact aren't the same type of objects which are explicitly like Account(parent) and Contact(Child).. Here we are modifying the objects from one another which is not allowed. 

Instead we need to have the check of getSobjectType() and further compare name and based on it we need to mention the if condition
